File: api/routers/track.py
from fastapi import APIRouter, Depends, BackgroundTasks
from models.schemas import TrackRequest, TrackResponse
from services.auth import verify_api_key
from services.database import get_supabase
from services.evaluator import run_evaluation

router = APIRouter()


@router.post("/track", response_model=TrackResponse)
async def track(
    body: TrackRequest,
    background_tasks: BackgroundTasks,
    api_key_info: dict = Depends(verify_api_key),
):
    supabase = get_supabase()

    result = (
        supabase.table("evaluations")
        .insert(
            {
                "api_key_id": api_key_info["id"],
                "question": body.question,
                "answer": body.answer,
                "chunks": [c.model_dump() for c in (body.chunks or [])],
                "latency_ms": body.latency_ms,
            }
        )
        .execute()
    )

    evaluation_id = result.data[0]["id"]

    # Groq評価はバックグラウンドで実行（10秒タイムアウト対策）
    background_tasks.add_task(
        run_evaluation,
        evaluation_id,
        body.question,
        body.answer,
        [c.model_dump() for c in (body.chunks or [])],
    )

    return TrackResponse(id=evaluation_id)

File: api-worker/src/routes/track.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { HTTPException } from "hono/http-exception";
import type { AppVariables, Bindings } from "../types";
import { verifyApiKey } from "../lib/auth";
import { runEvaluation } from "../lib/evaluator";
import { TrackRequestSchema } from "../schemas";

export const trackRoute = new Hono<{
  Bindings: Bindings;
  Variables: AppVariables;
}>();

trackRoute.post(
  "/track",
  verifyApiKey,
  zValidator("json", TrackRequestSchema),
  async (c) => {
    const body = c.req.valid("json");
    const apiKeyInfo = c.get("apiKeyInfo");

    const evaluationId = crypto.randomUUID();
    const chunks = body.chunks ?? [];

    const result = await c.env.DB.prepare(
      `INSERT INTO evaluations
         (id, api_key_id, question, answer, chunks, latency_ms)
       VALUES (?, ?, ?, ?, ?, ?)`,
    )
      .bind(
        evaluationId,
        apiKeyInfo.id,
        body.question,
        body.answer,
        JSON.stringify(chunks),
        body.latency_ms ?? null,
      )
      .run();

    if (!result.success) {
      throw new HTTPException(500, {
        message: `Failed to insert evaluation: ${result.error ?? "unknown"}`,
      });
    }

    // Workers AI評価はバックグラウンドで実行
    c.executionCtx.waitUntil(
      runEvaluation(c.env, evaluationId, body.question, body.answer, chunks),
    );

    return c.json({ id: evaluationId, status: "ok" });
  },
);
